litellm/tests/test_token_counter.py

```diff
@@ -56,9 +55,48 @@ def test_token_counter_normal_plus_function_calling():
     except Exception as e:
         pytest.fail(f"An exception occurred - {str(e)}")
 
-
 # test_token_counter_normal_plus_function_calling()
 
+@pytest.mark.parametrize(
+    "message_count_pair",
+    MESSAGES_TEXT,
+)
+def test_token_counter_textonly(message_count_pair):
+    counted_tokens = token_counter(
+        model="gpt-35-turbo",
+        messages=[message_count_pair["message"]]
+    )
+    assert counted_tokens == message_count_pair["count"]
+
+@pytest.mark.parametrize(
+    "message_count_pair",
+    MESSAGES_WITH_IMAGES,
+)
+def test_token_counter_with_images(message_count_pair):
+    counted_tokens = token_counter(
+        model="gpt-4o",
+        messages=[message_count_pair["message"]]
+    )
+    assert counted_tokens == message_count_pair["count"]
+
+
+@pytest.mark.parametrize(
+    "message_count_pair",
+    MESSAGES_WITH_TOOLS,
+)
+def test_token_counter_with_tools(message_count_pair):
+    counted_tokens = token_counter(
+        model="gpt-35-turbo",
+        messages=[message_count_pair["system_message"]],
+        tools=message_count_pair["tools"],
+        tool_choice=message_count_pair["tool_choice"],
+    )
+    expected_tokens = message_count_pair["count"]
+    diff = counted_tokens - expected_tokens
+    assert (
+        diff >= 0 and diff <= 3
+    ), f"Expected {expected_tokens} tokens, got {counted_tokens}. Counted tokens is only allowed to be off by 3 in the over-counting direction."
```

Comment:
Note that this doesnt do an exact comparison, as I haven't figured out how to get the counts to match exactly, and I don't know if anyone has, since this is all based on reverse-engineering. So I allow over-counting by 3.